### PR TITLE
[Merged by Bors] - add a debug label to storage buffers

### PR DESCRIPTION
**Pull request by brandon-reinhart**
# Objective

- Expose the wgpu debug label on storage buffer types.

## Solution

🐄

- Add an optional cow static string and pass that to the label field of create_buffer_with_data
- This pattern is already used by Bevy for debug tags on bind group and layout descriptors.

---

Example Usage:

A buffer is given a label using the label function. Alternatively a buffer may be labeled when it is created if the default() convention is not used.
ray_buf

Here is the buffer appearing with the correct name in RenderDoc. Previously the buffer would have an anonymous name such as "Buffer223":
buffer_named

### CODE REVIEW

File: crates/bevy_render/src/render_resource/storage_buffer.rs
Comment (alice-i-cecile) [inline crates/bevy_render/src/render_resource/storage_buffer.rs:68]: IMO this should be `set_label`; we need a seperate `label` getter function here (and below) as the fields aren't public.

Comment (alice-i-cecile) [inline crates/bevy_render/src/render_resource/storage_buffer.rs:68]: We should probably use a `impl Trait` argument here here to make the ergonomics a bit better.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/render_resource/storage_buffer.rs
Comment (brandon-reinhart) [reply crates/bevy_render/src/render_resource/storage_buffer.rs:68]: Is there any reason why label can't just be public? I don't really like trivial setters / getters, but I don't know you do in Rome.

Comment (brandon-reinhart) [reply crates/bevy_render/src/render_resource/storage_buffer.rs:68]: Ah, nevermind. If we do your second suggestion the setter would no longer be trivial.

Comment (alice-i-cecile) [reply crates/bevy_render/src/render_resource/storage_buffer.rs:68]: My first read is that this could just be public, but I don't know this code super well. I'll defer to @bzm3r and @superdump for their opinion here.

Comment (alice-i-cecile) [reply crates/bevy_render/src/render_resource/storage_buffer.rs:68]: If we can get away with it, I'd also prefer it to be pub.

Comment (superdump) [inline crates/bevy_render/src/render_resource/storage_buffer.rs:68]: I suppose if the label is changed, then the only way that would become visible would be for the buffer to be reallocated. That's a shame, but it's the reality. As such, I think we need to implement `get_label()`, and `set_label()`. In `set_label()`, check if the label changed, and if so, mark that for the next `reserve()` call, the buffer will need to be reallocated regardless of capacity. Then when reserve() is called, if the label changed bool is set, reset it to false, and reallocate the buffer with the new label. I'm suggesting doing it this way so we don't reallocate the buffer at the wrong time and break everything.

Also, I'm pretty sure `wgpu` accepts `Some(&str)` as we use things like `Some("main_opaque_pass_3d")` and that's a much nicer API than `Cow`.

**Review by superdump — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/render_resource/storage_buffer.rs
Comment (brandon-reinhart) [reply crates/bevy_render/src/render_resource/storage_buffer.rs]: Is there a reason ComputePipelineDescriptor stores its label as a Cow<'static, str> ?

**Comment by brandon-reinhart** [comment]
This is ready for a second look. Would be nice to get it in before 0.8.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_render/src/render_resource/storage_buffer.rs
Comment (alice-i-cecile) [reply crates/bevy_render/src/render_resource/storage_buffer.rs]: I suspect this was a misguided performance optimization. The `Cow` struct doesn't work the way that @cart thought it did originally.

**Comment by superdump** [comment]
This looks good to me. Would you mind doing the same for (Dynamic)UniformBuffer and BufferVec while you're at it? :)

**Comment by brandon-reinhart** [comment]
Added to uniform buffer (the commit comment says the wrong thing).

**Comment by superdump** [comment]
> Added to uniform buffer (the commit comment says the wrong thing).

And `BufferVec`? :) If you have the time.

**Comment by brandon-reinhart** [comment]
There you go. 

I notice that the BufferVec has a different API from the other buffers. It has a separate reserve and write_buffer, whereas the others only have write_buffer. Is there a reason for that difference?

**Comment by alice-i-cecile** [comment]
Somewhat off-topic for this thread, but I kind of want a trait to be able to abstract over all three of these and make sure the APIs stay synced 😆

**Review by alice-i-cecile — APPROVED** [review]

**Review by superdump — APPROVED** [review]
Just needs updating with main and fixing merge conflicts and it looks good to me.

**Comment by superdump** [comment]
> There you go.
> 
> I notice that the BufferVec has a different API from the other buffers. It has a separate reserve and write_buffer, whereas the others only have write_buffer. Is there a reason for that difference?

Legacy. :) Usage and patterns changed a bit so changes were made but not to all I guess.

**Comment by alice-i-cecile** [comment]
Once this is formatted I'll merge it in :)

**Pull request closed without merging**
